// picgo/dist/lib/Logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
class Logger {
    constructor(ctx) {
        this.level = {
            success: 'green',
            info: 'blue',
            warn: 'yellow',
            error: 'red'
        };
        this.ctx = ctx;
    }
    handleLog(type, msg) {
        if (!this.ctx.config.silent) {
            let log = chalk_1.default[this.level[type]](`[PicGo ${type.toUpperCase()}]: `);
            log += msg;
            console.log(log);
            return msg;
        }
        else {
            return;
        }
    }
    success(msg) {
        return this.handleLog('success', msg);
    }
    info(msg) {
        return this.handleLog('info', msg);
    }
    error(msg) {
        return this.handleLog('error', msg);
    }
    warn(msg) {
        return this.handleLog('warn', msg);
    }
}
exports.default = Logger;
